File: app/src/main/java/info/sanaebadi/tddandroidunittest/ui/noteslist/NotesListViewModel.java
package info.sanaebadi.tddandroidunittest.ui.noteslist;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MediatorLiveData;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModel;


import java.util.List;

import javax.inject.Inject;

import info.sanaebadi.tddandroidunittest.models.Note;
import info.sanaebadi.tddandroidunittest.repository.NoteRepository;
import info.sanaebadi.tddandroidunittest.ui.Resource;

public class NotesListViewModel extends ViewModel {

    private static final String TAG = "NotesListViewModel";

    // inject
    private final NoteRepository noteRepository;

    private MediatorLiveData<List<Note>> notes = new MediatorLiveData<>();

    @Inject
    public NotesListViewModel(NoteRepository noteRepository) {
        this.noteRepository = noteRepository;
    }

    public LiveData<Resource<Integer>> deleteNote(final Note note) throws Exception{
        return noteRepository.deleteNote(note);
    }

    public LiveData<List<Note>> observeNotes(){
        return notes;
    }

    public void getNotes(){
        final LiveData<List<Note>> source = noteRepository.getNotes();
        notes.addSource(source, new Observer<List<Note>>() {
            @Override
            public void onChanged(List<Note> notesList) {
                if(notesList != null){
                    notes.setValue(notesList);
                }
                notes.removeSource(source);
            }
        });
    }
}

File: app/src/main/java/info/sanaebadi/tddandroidunittest/ui/noteslist/NotesListViewModel.kt
package info.sanaebadi.tddandroidunittest.ui.noteslist

import androidx.lifecycle.LiveData
import androidx.lifecycle.MediatorLiveData
import androidx.lifecycle.ViewModel
import info.sanaebadi.tddandroidunittest.models.Note
import info.sanaebadi.tddandroidunittest.repository.NoteRepository
import info.sanaebadi.tddandroidunittest.ui.Resource
import javax.inject.Inject

class NotesListViewModel @Inject constructor(// inject
    private val noteRepository: NoteRepository
) : ViewModel() {
    private val notes =
        MediatorLiveData<List<Note>>()

    @Throws(Exception::class)
    fun deleteNote(note: Note?): LiveData<Resource<Int>> {
        return noteRepository.deleteNote(note)
    }

    fun observeNotes(): LiveData<List<Note>> {
        return notes
    }

    fun getNotes() {
        val source =
            noteRepository.notes
        notes.addSource(
            source
        ) { notesList ->
            if (notesList != null) {
                notes.value = notesList
            }
            notes.removeSource(
                source
            )
        }
    }

    companion object {
        private const val TAG = "NotesListViewModel"
    }

}